### PR TITLE
Lifecycle updates

### LINKED ISSUE #14714: Processor lifecycle changes
- Add explicit Close methods taking a CloseMode enum that currently just has one value (Detach). In the future, this list will grow as the service adds support for Drain and durable terminus. This should be added to Sender/Receiver/Processor/SessionProcessor.
- Implement IAsyncDisposable for consistency with sender/receiver.
- StopProcessing should not actually close links - it should stop receiving messages though. It should continue to wait for in-flight processing to complete.
- DisposeAsync should call CloseAsync with the default CloseMode of Detach.

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Fixes https://github.com/Azure/azure-sdk-for-net/issues/14714

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs [hunk 1]
```diff
@@ -32,13 +32,13 @@ public class ServiceBusClient : IAsyncDisposable
...
-        ///   Indicates whether or not this <see cref="ServiceBusClient"/> has been disposed.
+        ///   Indicates whether or not this <see cref="ServiceBusClient"/> has been closed.
         /// </summary>
         ///
         /// <value>
-        ///   <c>true</c> if the client is disposed; otherwise, <c>false</c>.
+        ///   <c>true</c> if the client is closed; otherwise, <c>false</c>.
         /// </value>
-        public bool IsDisposed { get; private set; } = false;
+        public bool IsClosed { get; private set; } = false;
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Client/ServiceBusClient.cs:41]: You may want to consider moving this out to use a `volatile` backing field.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs [hunk 2]
```diff
@@ -145,6 +145,15 @@ public class ServiceBusProcessor
...
         internal int MaxConcurrentCallsPerSession { get; }
 
+        /// <summary>
+        ///   Indicates whether or not this <see cref="ServiceBusProcessor"/> has been closed.
+        /// </summary>
+        ///
+        /// <value>
+        ///   <c>true</c> if the processor is closed; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsClosed { get; private set; } = false;
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs:155]: You may want to consider moving this out to use a `volatile` backing field.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs [hunk 3]
```diff
@@ -126,6 +138,7 @@ protected ServiceBusSessionProcessor()
         {
             add
             {
+                var msg = new ServiceBusMessage("sd");
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs]: Was this intentional?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs [hunk 4]
```diff
@@ -177,7 +183,7 @@ private async Task CreateReceiver(CancellationToken processorCancellationToken)
                     return;
                 }
                 _threadCount--;
-                if (_threadCount == 0)
+                if (_threadCount == 0 && !processorCancellationToken.IsCancellationRequested)
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs:186]: Does this need to be thread safe?  You may want to move it to an interlocked call.

**Review by jsquire — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs]: Nope, fixed.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/SessionReceiverManager.cs:186]: This is within a lock.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs [hunk 5]
```diff
@@ -19,7 +22,7 @@ namespace Azure.Messaging.ServiceBus
     /// property. The error handler is specified with the <see cref="ProcessErrorAsync"/> property.
     /// To start processing after the handlers have been specified, call <see cref="StartProcessingAsync"/>.
     /// </summary>
-    public class ServiceBusSessionProcessor
+    public class ServiceBusSessionProcessor : IAsyncDisposable
```
Comment (MiYanni) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs:25]: Is it a breaking change to make a type go from non-disposable to disposable? Mainly, programs that were using this prior would not be calling `.Dispose()`. Thus, the 'usage contract' has changed (if that makes sense).

**Review by MiYanni — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusSessionProcessor.cs:25]: I think as long as you make sure that old code could work correctly without calling Dispose, it wouldn't necessarily be breaking. But this is a preview so it should be okay 😄 

https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#types

**Pull request merged**